pylint/checkers/refactoring/recommendation_checker.py

```diff
@@ -83,21 +83,23 @@ def _check_consider_iterating_dictionary(self, node: nodes.Call) -> None:
             return
         if node.func.attrname != "keys":
             return
-        if (
-            isinstance(node.parent, (nodes.For, nodes.Comprehension))
-            or isinstance(node.parent, nodes.Compare)
-            and any(
-                op
-                for op, comparator in node.parent.ops
-                if op == "in" and comparator is node
-            )
+        inferred = utils.safe_infer(node.func)
```

Comment:
I think it is better if we do the (performance-heavy) inference after we have done the check for ancestors (like we do in the current code).

With the changes in the other comment this should be easily doable.

Reply:
fixed this too